ngraph/core/src/op/matmul.cpp

```diff
@@ -113,68 +53,162 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
-            NGRAPH_CHECK(arg0_shape_update[0] == arg1_shape_update[arg1_rank - 2],
-                         "Incompatible arg shapes");
-            arg1_shape_update.erase(arg1_shape_update.begin() + arg1_rank - 2);
-            output_shape = arg1_shape_update;
+            // If the second input is 1D tensor, it is unsqueezed to 2D tensor (column vector)
+            // by adding axes with size 1 at COL_INDEX_DIM, to the right of the shape.
+            // For example {S} will be reshaped to {S, 1}.
+            arg1_shape_tmp.insert(arg1_shape_tmp.end(), 1);
+            arg1_rank = arg1_shape_tmp.size();
         }
-        else if (arg1_rank == 1)
+
+        // Check matrices dimensions compatibility,
+        // COL_INDEX_DIM of the first matrix has to match ROW_INDEX_DIM of the second matrix.
+        // Error is not thrown for dynamic dimensions bounds without intersection
+        // to ensure MatMul backward compatibility.
+        auto merged_dimension = Dimension::dynamic();
+        auto arg0_col_dim = arg0_shape_tmp[arg0_rank - 1];
+        auto arg1_row_dim = arg1_shape_tmp[arg1_rank - 2];
+        NGRAPH_CHECK(Dimension::merge(merged_dimension, arg0_col_dim, arg1_row_dim) ||
+                         arg0_col_dim.is_dynamic() || arg1_row_dim.is_dynamic(),
+                     "Incompatible MatMul matrix dimension. ",
+                     "First input dimension=",
+                     arg0_col_dim,
+                     " at COL_INDEX_DIM=",
+                     (arg0_rank - 1),
+                     " doesn't match the second input dimension=",
+                     arg1_row_dim,
+                     " at ROW_INDEX_DIM=",
+                     (arg1_rank - 2));
+
+        // 3. If ranks of input arguments are different after steps 1 and 2,
+        // the smaller tensor is unsqueezed from the left side of the shape
+        // by necessary number of axes to make both shapes of the same rank.
+        if (arg0_rank != arg1_rank)
         {
-            // i.e., arg0 shape {2, 2, 3}, arg1 shape{3}, output shape {2, 2}
-            NGRAPH_CHECK(arg1_shape_update[0] == arg0_shape_update[arg0_rank - 1],
-                         "Incompatible arg shapes");
-            arg0_shape_update.erase(arg0_shape_update.begin() + arg0_rank - 1);
-            output_shape = arg0_shape_update;
+            // Expand small_batch_matrix (with 1) to have the same rank as big_batch_matrix
+            // size_t delta_rank = big_batch_matrix.size() - small_batch_matrix.size();
+            // small_batch_matrix.insert(small_batch_matrix.begin(), delta_rank, 1);
+            int64_t delta_rank = arg0_rank - arg1_rank;
+            delta_rank < 0 ? arg0_shape_tmp.insert(arg0_shape_tmp.begin(), std::abs(delta_rank), 1)
+                           : arg1_shape_tmp.insert(arg1_shape_tmp.begin(), std::abs(delta_rank), 1);
+            arg0_rank = arg0_shape_tmp.size();
+            arg1_rank = arg1_shape_tmp.size();
```

Comment:
arg1_rank is never used below. Ternary operator is used instead of if/else statement (why?). Comment is not relevant to the code. OK, let me just propose this short form without all those `abs` etc.:

```
if (arg0_rank < arg1_rank)
    arg0_shape_tmp.insert(arg0_shape_tmp.begin(), arg1_rank - arg0_rank, 1);
else if (arg0_rank > arg1_rank)
    arg1_shape_tmp.insert(arg1_shape_tmp.begin(), arg0_rank - arg1_rank, 1);
auto max_rank = arg0_shape_tmp.size();   // both arg0_shape_tmp and arg1_shape_tmp have identical size now
```

Reply:
Thanks, done